api/include/opentelemetry/context/context.h

```diff
@@ -0,0 +1,145 @@
+#include "opentelemetry/trace/key_value_iterable_view.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace context
+{
+
+// The context class provides a context identifier. Is built as a linked list
+// of DataList nodes and each context holds a shared_ptr to a place within
+// the list that determines which keys and values it has access to. All that
+// come before and none that come after.
+class Context
+{
+
+public:
+  // Creates a context object from a map of keys and identifiers, this will
+  // hold a shared_ptr to the head of the DataList linked list
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context(const T &keys_and_values)
+  {
+    head_ = nostd::shared_ptr<DataList>{new DataList(keys_and_values)};
+  }
+
+  // Creates a context object from a key and value, this will
+  // hold a shared_ptr to the head of the DataList linked list
+  Context(nostd::string_view key, ContextValue value)
+  {
+    head_ = nostd::shared_ptr<DataList>{new DataList(key, value)};
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data. It attaches the
+  // exisiting list to the end of the new list.
+  template <class T, nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+  Context SetValues(T &values) noexcept
+  {
+    Context context = Context(values);
+    nostd::shared_ptr<DataList> last_node;
+    for (nostd::shared_ptr<DataList> data = context.head_; data != nullptr; data = data->next_)
+    {
+      last_node = data;
+    }
+    last_node->next_ = head_;
+    return context;
+  }
+
+  // Accepts a new iterable and then returns a new context that
+  // contains the new key and value data. It attaches the
+  // exisiting list to the end of the new list.
+  Context SetValue(nostd::string_view key, ContextValue value) noexcept
+  {
+    Context context      = Context(key, value);
+    context.head_->next_ = head_;
+    return context;
+  }
+
+  // Returns the value associated with the passed in key.
+  context::ContextValue GetValue(nostd::string_view key)
+  {
+    for (nostd::shared_ptr<DataList> data = head_; data != nullptr; data = data->next_)
+    {
+      if (key.size() == data->key_length_)
+      {
+        if (memcmp(key.data(), data->key_, data->key_length_) == 0)
+        {
+          return data->value_;
+        }
+      }
+    }
+    return "";
+  }
+
+private:
+  Context() = default;
+
+  // A linked list to contain the keys and values of this context node
+  class DataList
+  {
+  public:
+    nostd::shared_ptr<DataList> next_;
+
+    char *key_;
+
+    size_t key_length_;
+
+    ContextValue value_;
+
+    DataList() { next_ = nullptr; }
+
+    // Builds a data list off of a key and value iterable and returns the head
+    template <class T,
+              nostd::enable_if_t<trace::detail::is_key_value_iterable<T>::value> * = nullptr>
+    DataList(const T &keys_and_vals)
```

Comment:
Please add a test for initializing this with an empty map (for `keys_and_vals`). I think this will fail.

Reply:
Added test and fixed functionality.

Reply:
Please add an additional move constructor:
```c++
DataList &operator=(DataList &&other)
{
  key_length_ = other.key_length_;
  value_      = std::move(other.value_);
  next_       = std::move(other.next_);

  key_       = other.key_;
  other.key_ = nullptr;

  return *this;
}
```

Then you can simplify this constructor and avoid code duplication:
```c++
template <class T>
DataList(const T &keys_and_vals) : key_{nullptr}
{
  bool first = true;
  auto *node = this;
  for (auto &iter : keys_and_vals)
  {
    if (first)
    {
      *node = std::move(DataList(iter.first, iter.second));
      first = false;
    } 
    else
    {
      node->next_ = nostd::shared_ptr<DataList>(new DataList(iter.first, iter.second));
      node        = node->next_.get();
    } 
  } 
}